Suggest checking out reference branch before build when reference branch is not found

To keep #390  clean I am creating a new issue to address the error message that I am getting.

For people that are new to a project using GIB, it might be helpful to change the following message 

```
[ERROR] Exception during gitflow-incremental-builder execution occurred. Git branch of name 'main-reference' not found. -> [Help 1]
```

Into 

```
[ERROR] Exception during gitflow-incremental-builder execution occurred. Git branch of name 'main-reference' not found. Consider checking out the branch first with git checkout main-reference, then checking out the branch you tried to build (git checkout <CURRENT_BRANCH_NAME>. -> [Help 1].
```

--- a/src/main/java/com/vackosar/gitflowincrementalbuild/boundary/MavenLifecycleParticipant.java
+++ b/src/main/java/com/vackosar/gitflowincrementalbuild/boundary/MavenLifecycleParticipant.java
@@ -91,7 +91,7 @@ public class MavenLifecycleParticipant extends AbstractMavenLifecycleParticipant
                 logger.info("gitflow-incremental-builder execution skipped: {}", (isSkipExecException ? e.getMessage() : e.toString()));
                 logger.debug("Full exception:", e);
             } else {
-                throw new MavenExecutionException("Exception during gitflow-incremental-builder execution occurred.", e);
+                throw new MavenExecutionException("Failed to execute gitflow-incremental-builder.", e);
             }
         }
         logger.info("gitflow-incremental-builder exiting...");

--- a/src/main/java/com/vackosar/gitflowincrementalbuild/control/DifferentFiles.java
+++ b/src/main/java/com/vackosar/gitflowincrementalbuild/control/DifferentFiles.java
@@ -99,39 +99,38 @@ public class DifferentFiles {
         }
 
         private Set<Path> getBranchDiff() throws IOException {
-            RevCommit base = getBranchCommit(configuration.baseBranch);
-            final TreeWalk treeWalk = new TreeWalk(git.getRepository());
-            try {
+            RevCommit base = getBranchCommit(configuration.baseBranch, false);
+            try (final TreeWalk treeWalk = new TreeWalk(git.getRepository())) {
                 treeWalk.addTree(base.getTree());
                 treeWalk.addTree(resolveReference(base).getTree());
                 treeWalk.setFilter(TreeFilter.ANY_DIFF);
                 treeWalk.setRecursive(true);
                 return getDiff(treeWalk, workTree);
-            } finally {
-                treeWalk.close();
             }
         }
 
         private void checkout() throws IOException, GitAPIException {
             if (! (HEAD.equals(configuration.baseBranch) || configuration.baseBranch.startsWith("worktrees/")) && ! git.getRepository().getFullBranch().equals(configuration.baseBranch)) {
-                logger.info("Checking out base branch " + configuration.baseBranch + "...");
+                logger.info("Checking out base branch " + configuration.baseBranch);
                 git.checkout().setName(configuration.baseBranch).call();
             }
         }
 
         private void fetch() throws GitAPIException {
             if (!configuration.disableBranchComparison && configuration.fetchReferenceBranch) {
-                fetch(configuration.referenceBranch);
+                fetch(configuration.referenceBranch, true);
             }
             if (configuration.fetchBaseBranch) {
-                fetch(configuration.baseBranch);
+                fetch(configuration.baseBranch, false);
             }
         }
 
-        private void fetch(String branchName) throws GitAPIException {
+        private void fetch(String branchName, boolean reference) throws GitAPIException {
             logger.info("Fetching branch " + branchName);
             if (!branchName.startsWith(REFS_REMOTES)) {
-                throw new IllegalArgumentException("Branch name '" + branchName + "' is not tracking branch name since it does not start " + REFS_REMOTES);
+                throw new IllegalArgumentException("Cannot fetch local " + (reference ? "reference" : "base") + " branch '" + branchName + "'. "
+                        + "Only remote tracking branches can be fetched, meaning branches starting with '" + REFS_REMOTES + "'. "
+                        + "Make sure to not confuse remote tracking branches with local branches, 'git branch -a' is your friend!");
             }
             String remoteName = extractRemoteName(branchName);
             String shortName = extractShortName(remoteName, branchName);
@@ -158,14 +157,15 @@ public class DifferentFiles {
         }
 
         private RevCommit getMergeBase(RevCommit baseCommit, RevCommit referenceHeadCommit) throws IOException {
-            RevWalk walk = new RevWalk(git.getRepository());
-            walk.setRevFilter(RevFilter.MERGE_BASE);
-            walk.markStart(walk.lookupCommit(baseCommit));
-            walk.markStart(walk.lookupCommit(referenceHeadCommit));
-            RevCommit commit = walk.next();
-            walk.close();
-            logger.info("Using merge base of id: " + commit.getId());
-            return commit;
+            try (final RevWalk walk = new RevWalk(git.getRepository())) {
+                walk.setRevFilter(RevFilter.MERGE_BASE);
+                walk.markStart(walk.lookupCommit(baseCommit));
+                walk.markStart(walk.lookupCommit(referenceHeadCommit));
+                RevCommit commit = walk.next();
+                walk.close();
+                logger.info("Using merge base of id: " + commit.getId());
+                return commit;
+            }
         }
 
         private Set<Path> getDiff(TreeWalk treeWalk, Path gitDir) throws IOException {
@@ -179,26 +179,34 @@ public class DifferentFiles {
             return paths;
         }
 
-        private RevCommit getBranchCommit(String branchName) throws IOException {
+        private RevCommit getBranchCommit(String branchName, boolean reference) throws IOException {
             Repository repository = git.getRepository();
             ObjectId objectId = repository.resolve(branchName);
 
+            boolean isRemoteTrackingBranch = branchName.startsWith(REFS_REMOTES);
+            String branchDesc = String.format("%s %s branch '%s'",
+                    isRemoteTrackingBranch ? "remote tracking" : "local",
+                    reference ? "reference" : "base",
+                    branchName);
             if (objectId == null) {
                 if (repository.simplify(branchName) != null) {
-                    throw new SkipExecutionException("Could not get Git ObjectId for branch of name '" + branchName
-                            + "'. Is this repository empty (no commits yet)?");
+                    throw new SkipExecutionException("Could not get Git ObjectId for " + branchDesc
+                            + ". Is this repository empty (no commits yet)?");
                 }
-                if (branchName.startsWith(REFS_REMOTES) && repository.getRemoteNames().isEmpty()) {
-                    throw new SkipExecutionException("Could not get Git ObjectId for branch of name '" + branchName
-                            + "'. No remotes found at all, push still pending?");
+                if (isRemoteTrackingBranch && repository.getRemoteNames().isEmpty()) {
+                    throw new SkipExecutionException("Could not get Git ObjectId for " + branchDesc
+                            + ". No remotes found at all, push still pending?");
                 }
-                throw new IllegalArgumentException("Git branch of name '" + branchName + "' not found.");
+                throw new IllegalArgumentException("Git " + branchDesc + " not found. "
+                        + "Make sure it exists by creating it explicitly via 'git branch/checkout/switch ...' "
+                        + "or by fetching it from the remote repository via 'git fetch <remote> <branch>:<branch>' (or just 'git fetch <branch>'). "
+                        + "Also make sure to not confuse remote tracking branches (refs/remotes/...) with local branches, 'git branch -a' is your friend!");
             }
-            final RevWalk walk = new RevWalk(repository);
-            RevCommit commit = walk.parseCommit(objectId);
-            walk.close();
-            logger.info("Reference commit of branch " + branchName + " is commit of id: " + commit.getId());
-            return commit;
+            try (final RevWalk walk = new RevWalk(repository)) {
+                RevCommit commit = walk.parseCommit(objectId);
+                logger.info("Reference commit of " + branchDesc + " has id: " + commit.getId());
+                return commit;
+            }
         }
 
         private Set<Path> getChangesFromStatus() throws GitAPIException {
@@ -219,7 +227,7 @@ public class DifferentFiles {
         }
 
         private RevCommit resolveReference(RevCommit base) throws IOException {
-            RevCommit refHead = getBranchCommit(configuration.referenceBranch);
+            RevCommit refHead = getBranchCommit(configuration.referenceBranch, true);
             if (configuration.compareToMergeBase) {
                 return getMergeBase(base, refHead);
             } else {

--- a/src/test/java/com/vackosar/gitflowincrementalbuild/control/BaseDifferentFilesTest.java
+++ b/src/test/java/com/vackosar/gitflowincrementalbuild/control/BaseDifferentFilesTest.java
@@ -87,6 +87,17 @@ public abstract class BaseDifferentFilesTest extends BaseRepoTest {
         assertThat(git.log().setMaxCount(1).call().iterator().next().getFullMessage()).isEqualTo(message);
     }
 
+    protected void assertFetchFileCommitExistsInDevelop() throws Exception {
+        assertFetchFileCommitExistsInDevelop(true);
+    }
+
+    protected void assertFetchFileCommitExistsInDevelop(boolean remoteDevelop) throws Exception {
+        Git localGit = localRepoMock.getGit();
+        localGit.reset().setMode(ResetCommand.ResetType.HARD).call();
+        localGit.checkout().setName(remoteDevelop ? REMOTE_DEVELOP : DEVELOP).call();
+        assertCommitExists(FETCH_FILE, localGit);
+    }
+
     protected Set<Path> invokeUnderTest() throws Exception {
         return invokeUnderTest(getMavenSessionMock());
     }

--- a/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesHttpFetchBasicAuthTest.java
+++ b/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesHttpFetchBasicAuthTest.java
@@ -5,8 +5,6 @@ import static com.vackosar.gitflowincrementalbuild.mocks.server.TestServerType.H
 import java.io.IOException;
 import java.nio.file.Files;
 
-import org.eclipse.jgit.api.Git;
-import org.eclipse.jgit.api.ResetCommand;
 import org.eclipse.jgit.lib.StoredConfig;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -39,10 +37,7 @@ public class DifferentFilesHttpFetchBasicAuthTest extends BaseDifferentFilesTest
 
         invokeUnderTest();
 
-        Git localGit = localRepoMock.getGit();
-        localGit.reset().setMode(ResetCommand.ResetType.HARD).call();
-        localGit.checkout().setName(REMOTE_DEVELOP).call();
-        assertCommitExists(FETCH_FILE, localGit);
+        assertFetchFileCommitExistsInDevelop();
     }
 
     private String buildCredentialsFileContent() {

--- a/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesHttpFetchTest.java
+++ b/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesHttpFetchTest.java
@@ -1,7 +1,6 @@
 package com.vackosar.gitflowincrementalbuild.control;
 
-import org.eclipse.jgit.api.Git;
-import org.eclipse.jgit.api.ResetCommand;
+import org.assertj.core.api.Assertions;
 import org.junit.jupiter.api.Test;
 
 import com.vackosar.gitflowincrementalbuild.mocks.server.TestServerType;
@@ -20,9 +19,17 @@ public class DifferentFilesHttpFetchTest extends BaseDifferentFilesTest {
 
         invokeUnderTest();
 
-        Git localGit = localRepoMock.getGit();
-        localGit.reset().setMode(ResetCommand.ResetType.HARD).call();
-        localGit.checkout().setName(REMOTE_DEVELOP).call();
-        assertCommitExists(FETCH_FILE, localGit);
+        assertFetchFileCommitExistsInDevelop();
+    }
+
+    @Test
+    public void fetch_invalidLocalBranch() throws Exception {
+        addCommitToRemoteRepo(FETCH_FILE);
+        projectProperties.setProperty(Property.fetchReferenceBranch.prefixedName(), "true");
+        projectProperties.setProperty(Property.referenceBranch.prefixedName(), "develop");
+
+        Assertions.assertThatThrownBy(this::invokeUnderTest)
+                .isExactlyInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("Cannot fetch local reference branch 'develop'");
     }
 }

--- a/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesSshFetchTest.java
+++ b/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesSshFetchTest.java
@@ -15,8 +15,6 @@ import java.util.Collections;
 import java.util.HashSet;
 
 import org.apache.commons.lang3.SystemUtils;
-import org.eclipse.jgit.api.Git;
-import org.eclipse.jgit.api.ResetCommand;
 import org.eclipse.jgit.transport.SshSessionFactory;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Assumptions;
@@ -122,10 +120,7 @@ public class DifferentFilesSshFetchTest extends BaseDifferentFilesTest {
 
         invokeUnderTest();
 
-        Git localGit = localRepoMock.getGit();
-        localGit.reset().setMode(ResetCommand.ResetType.HARD).call();
-        localGit.checkout().setName(REMOTE_DEVELOP).call();
-        assertCommitExists(FETCH_FILE, localGit);
+        assertFetchFileCommitExistsInDevelop();
     }
 
     private Path writePrivateKey(String fileName, String key) throws IOException {

--- a/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesTest.java
+++ b/src/test/java/com/vackosar/gitflowincrementalbuild/control/DifferentFilesTest.java
@@ -210,10 +210,7 @@ public class DifferentFilesTest extends BaseDifferentFilesTest {
 
         invokeUnderTest();
 
-        Git localGit = localRepoMock.getGit();
-        localGit.reset().setMode(ResetCommand.ResetType.HARD).call();
-        localGit.checkout().setName(REMOTE_DEVELOP).call();
-        assertCommitExists(FETCH_FILE, localGit);
+        assertFetchFileCommitExistsInDevelop();
     }
 
     @Test
@@ -227,9 +224,7 @@ public class DifferentFilesTest extends BaseDifferentFilesTest {
 
         invokeUnderTest();
 
-        localGit.reset().setMode(ResetCommand.ResetType.HARD).call();
-        localGit.checkout().setName(REMOTE_DEVELOP).call();
-        assertCommitExists(FETCH_FILE, localGit);
+        assertFetchFileCommitExistsInDevelop();
     }
 
     @Test
@@ -239,7 +234,7 @@ public class DifferentFilesTest extends BaseDifferentFilesTest {
         Throwable thrown = catchThrowable(() -> invokeUnderTest());
 
         assertThat(thrown).isInstanceOf(IllegalArgumentException.class);
-        assertThat(thrown).hasMessage("Git branch of name 'FOO' not found.");
+        assertThat(thrown).hasMessageStartingWith("Git local reference branch 'FOO' not found.");
     }
 
     @Test

